src/metabase/util/queue.clj

```diff
@@ -63,24 +70,155 @@
+  (contains? @listeners listener-name))
+
+(mu/defn- listener-thread [listener-name :- :string
+                           queue :- (ms/InstanceOfClass BlockingQueue)
+                           handler :- [:=> [:cat [:sequential :any]] :any]
+                           {:keys [success-handler err-handler max-batch-messages max-next-ms]} :- ::listener-options]
+  (log/debugf "Thread for listener %s started" listener-name)
+  (while true
+    (try
+      (log/debugf "Listener %s waiting for next batch..." listener-name)
+      (let [batch (take-batch! queue Long/MAX_VALUE max-batch-messages max-next-ms)]
+        (if (seq batch)
+          (do
+            (log/debugf "Listener %s processing batch of %d" listener-name (count batch))
+            (log/tracef "Listener %s processing batch: %s" listener-name batch)
+            (let [timer (u/start-timer)
+                  output (handler batch)
+                  duration (u/since-ms timer)]
+              (log/debugf "Listener %s processed batch in %.0fms" listener-name duration)
+              (success-handler output duration listener-name)))
+          (log/debugf "Listener %s found no items to process" listener-name)))
+      (catch InterruptedException e
+        (log/debugf "Listener thread %s stopped" listener-name)
+        (throw e))
+      (catch Exception e
+        (err-handler e listener-name)
+        (log/errorf e "Error in %s while processing batch" listener-name))))
+  (log/infof "Listener %s stopped" listener-name))
+
+(mu/defn listen!
+  "Starts an async listener on the given queue.
+
+  Arguments:
+  - listener-name: A unique string. Calls to register another listener with the same name will be a no-op
+  - queue: The queue to listen on
+  - handler: A function taking a list of 1 or more values that have been sent to the queue.
+
+  Options:
+  - success-handler: A function called when handler does not throw an exception. Accepts [result-of-handler, duration-in-ms, listener-name]
+  - err-handler: A function called when the handler throws an exception. Accepts [exception, duration-in-ms, listener-name]
+  - pool-size: Number of threads in the listener. Default: 1
+  - max-batch-messages: Max number of items to batch up before calling handler. Default 50
+  - max-next-ms: Max number of ms to wait for each additional message before calling the handler. Default 100"
+  [listener-name :- :string
+   queue :- (ms/InstanceOfClass BlockingQueue)
+   handler :- [:=> [:cat [:sequential :any]] :any]
+   {:keys [success-handler
+           err-handler
+           pool-size
+           max-batch-messages
+           max-next-ms]
+    :or   {success-handler (constantly nil)
+           err-handler (constantly nil)
+           pool-size       1
+           max-batch-messages 50
+           max-next-ms     100}} :- ::listener-options]
+  (if (listener-exists? listener-name)
+    (log/errorf "Listener %s already exists" listener-name)
+
+    (let [executor (Executors/newFixedThreadPool
```

Comment:
I really like [`claypoole`](https://github.com/clj-commons/claypoole), think it might fit this well, but it's mostly just a clojurey wrapper around this kind of thing.

Reply:
Claypoole is nicer, swiched to that